llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -5088,7 +5108,7 @@ InstructionCost LoopVectorizationCostModel::expectedCost(ElementCount VF) {
     // cost by the probability of executing it. blockNeedsPredication from
     // Legal is used so as to not include all blocks in tail folded loops.
     if (VF.isScalar() && Legal->blockNeedsPredication(BB))
-      BlockCost /= getPredBlockCostDivisor(CostKind);
+      BlockCost /= getPredBlockCostDivisor(CostKind, BB);
```

Comment:
Here the call to `getPredBlockCostDivisor` is already guarded by `Legal->blockNeedsPredication(BB)`. Instead of changing the meaning/behaviour of `getPredBlockCostDivisor` is it perhaps better to simply guard all calls to the function in a similar way? Then you can add a comment to `getPredBlockCostDivisor` saying that it should only ever be called for blocks that are predicated in the original scalar code.

Alternatively, you could remove the `Legal->blockNeedsPredication(BB)` check here and update the comment above `getPredBlockCostDivisor` saying that it will always return 1 for blocks that are predicated due to tail-folding.

Reply:
I've removed the redundant `Legal->blockNeedsPredication(BB)` in cf6b43532792f46850c158c54c9dc28c912b7148 and updated the comment. I figured we're going to need to plumb through the basic block anyway in #158690 so we might as well do it in this PR